### PR TITLE
pTSL/TSL conversion - command line interface

### PR DESCRIPTION
**Pull request by eyezhick**

### CODE REVIEW

**Comment by eyezhick** [comment]
Having a separate flag as (p)tsl_out_file is a bit messy, but I thought bombarding/overloading the general --output_file flag with more information was messier. Ideally, I would have it so the –to-(p)tsl flag can take arguments for filenames, but I couldn’t quite get it to do what I want because it needed to take 0,1, or 2 arguments. Not sure how I can differentiate no flag with 0 argument.

File: src/TSL.py [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+            <symbol n>
+
+            FACTORS_FILE
+            <factor 1>
+            <factor 2>
+            ...
+            <factor n>
+
+
+            <symbol i> is a single segment with no spaces. Each segment is a tier alphabet.
```
Comment (connormayer) [inline src/TSL.py]: I would reword this to say '<symbol i> is a member of the tier alphabet, and consists of one or more characters without spaces.'

File: src/TSL.py [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+            <factor i> is a string of segments separated by spaces.
```
Comment (connormayer) [inline src/TSL.py]: Maybe say 'a string of segments from the tier alphabet'

File: src/TSL.py [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+class TSL:
+    '''
+        tier is the tier alphabet, k-factors is the strictly k-local grammar
+        tier is a list e.g. [a, bc]
+        k-factors is a grammar list of tuple containing each char as element e.g. [(ph,a),(sh,b),(a,c)]
+    '''
+    def __init__(self, tier, k_factors):
+        self.tier = tier
+        self.k_factors = k_factors
+        self.k = max([len(f) for f in self.k_factors])
```
Comment (connormayer) [inline src/TSL.py]: We should probably add a check to the constructors here and in `pTSL.py` to require all k-factors to be the same length. It's kind of convenient not to have this requirement, but the `check_k_factors` method here and the `get_kfactors` method in `pTSL.py` will both behave incorrectly if this isn't the case (e.g., if the factors are a mix of 2- and 3-factors, it will only reject strings on the basis of the 3-factors).

File: src/TSL.py [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+    def eval_to_csv(corpus_data, results, filename=None):
+        '''
+            Save the result of eval to output filename
+        '''
+        if not filename:
+            filename = DEFAULT_EVAL_FILENAME
+        with open(filename, 'w', newline="") as f:
+            csv_writer = csv.writer(f)
+            for data, val in list(zip(corpus_data, results)):
+                csv_writer.writerow([' '.join(map(str, data)), round(val, 6)])
```
Comment (connormayer) [inline src/TSL.py]: The `val` in this case will be booleans won't they? So we shouldn't need the `round` bit.

File: src/TSL.py [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+                csv_writer.writerow(sym)
+
+    @staticmethod
+    def eval_to_csv(corpus_data, results, filename=None):
+        '''
+            Save the result of eval to output filename
+        '''
+        if not filename:
+            filename = DEFAULT_EVAL_FILENAME
+        with open(filename, 'w', newline="") as f:
```
Comment (connormayer) [inline src/TSL.py:158]: What's the deal with `newline=""` for all these csv files? It seems like the csv writer puts in the proper newline characters, but I don't understand what this is doing.

File: src/TSL.py [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+                corpus_filename: path to save corpus evaluations.
+        '''
+        corpus_data = self.read_corpus_file(corpus_file)
+        results = []
+        for string in corpus_data:
+            results.append(self.val(string))
+
+        self.eval_to_csv(corpus_data, results, corpus_filename)
+
+    def TSL_to_pTSL(self, tier_filename, factor_filename):
```
Comment (connormayer) [inline src/TSL.py]: You can just call this `to_pTSL`. The `TSL` part is implied since it's a method of the `TSL` class.

File: src/TSL.py [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+            Convert TSL grammar to pTSL and output a pTSL grammar by saving the tier and k-factor files.
+
+        '''
+        if not tier_filename:
+            filename = DEFAULT_PTSL_TIER_FILENAME
+        if not factor_filename:
+            filename = DEFAULT_PTSL_FACTOR_FILENAME
+        probs = {}
+
+        for tier in self.tier:
```
Comment (connormayer) [inline src/TSL.py]: I'd call the loop variable `symbol` or something like that. There's only one tier.

File: src/TSL.py [hunk 8]
```diff
@@ -0,0 +1,224 @@
...
+        factors = self.k_factors
+
+        with open(tier_filename, 'w', newline="") as f:
+            csv_writer = csv.writer(f)
+            for key, value in probs.items():
+                csv_writer.writerow([key, value])
+
+        with open(factor_filename, 'w', newline="") as f:
+            csv_writer = csv.writer(f)
+            for sym in factors:
```
Comment (connormayer) [inline src/TSL.py:148]: This ends up writing each factor as a comma separated row. Each symbol needs to be separated by a space. You can fix this by just setting `sep=' '` in the definition of the writer.

You should test this end-to-end by doing the following:

1. Define a TSL grammar and convert it to pTSL.
2. Create a pTSL grammar by reading the output from Step 1.
3. Convert the grammar created in Step 2 to a TSL grammar.
4. Create a TSL grammar by reading the output from Step 3.

This will make sure the input and output formats agree.

File: src/pTSL.py [hunk 9]
```diff
@@ -2,13 +2,17 @@
 import csv
 from scipy.optimize import minimize
 from numpy.random import rand
+import TSL
```
Comment (connormayer) [inline src/pTSL.py]: I don't think you use this anywhere.

File: src/pTSL.py [hunk 10]
```diff
@@ -185,6 +187,9 @@ def evaluate_proj(self, proj_probs, param, corpus_probs):
         return sse
 
     def train(self, corpus_file):
+        '''
+            Given the corpus file with corpus data and it's conditional probabilities, update the prob projections
```
Comment (connormayer) [inline src/pTSL.py]: `its`

Also, I would say `learn the optimal projection probabilities using maximum likelihood estimation.`

File: src/pTSL.py [hunk 11]
```diff
@@ -227,6 +232,54 @@ def probs_to_csv(self, filename=None):
             for key, value in self.probs.items():
                 csv_writer.writerow([key, value])
 
+    def pTSL_to_TSL(self, tier_filename=None, factor_filename=None):
```
Comment (connormayer) [inline src/pTSL.py]: Just `to_TSL` is fine

File: src/pTSL.py [hunk 12]
```diff
@@ -227,6 +232,54 @@ def probs_to_csv(self, filename=None):
...
+        '''
+            Convert pTSL grammar to TSL and output a TSL grammar by saving the tier and k-factor files.
+
+            Case 1:
+                If the k-factors in pTSL grammar contain only symbols with a projection prob < 1, then
+                conversion is trivial.
+            Case 2:
+                If the k-factors in pTSL grammar contain only symbols with a projection prob of 1, then
+                conversion is possible.
+
```
Comment (connormayer) [inline src/pTSL.py:246]: Otherwise what happens?

File: src/pTSL.py [hunk 13]
```diff
@@ -227,6 +232,54 @@ def probs_to_csv(self, filename=None):
...
+        if all([all([self.probs[sym] == 1 for sym in factor]) for factor in self.factors]):
+            tier = [x for x in self.probs if self.probs[x] > 0]
+            with open(tier_filename, 'w', newline="") as f:
+                csv_writer = csv.writer(f)
+                for sym in tier:
+                    csv_writer.writerow(sym)
+
+            with open(factor_filename, 'w', newline="") as f:
+                csv_writer = csv.writer(f)
+                for sym in self.factors:
```
Comment (connormayer) [inline src/pTSL.py:271]: I think this will also write each row of the factor file with comma separation rather than spaces.

File: src/pTSL.py [hunk 14]
```diff
@@ -277,7 +330,7 @@ def read_corpus_file(corpus_file, train=False):
     )
 
     parser.add_argument(
-        'input_file', type=str, action='store', nargs=3,
+        'input_file', type=str, action='store', nargs='*',
```
Comment (connormayer) [inline src/pTSL.py]: I think you can make this less awkward by splitting this into three separate positional arguments. The first two arguments correspond to the tier and factor files and have `nargs=1` and the third corresponds to the corpus and has `nargs='?'`. Then you just need to add a check that something has been passed in to the third argument if the `to_tsl` flag hasn't been set.

File: src/pTSL.py [hunk 15]
```diff
@@ -297,47 +350,65 @@ def read_corpus_file(corpus_file, train=False):
              'If verbose, pass another path to projections as the second argument.'
     )
 
+    parser.add_argument(
+        '--to_tsl', action='store_true',
+        help='Convert pTSL grammar to TSL grammar (if possible).'
+    )
+
+    parser.add_argument(
+        '--tsl_output_file', type=str, default=[None, None], nargs=2,
```
Comment (connormayer) [inline src/pTSL.py]: I'd call this `tsl_output_files` to make it clear there's more than one, or split it into two separate arguments.

File: src/pTSL.py [hunk 16]
```diff
@@ -297,47 +350,65 @@ def read_corpus_file(corpus_file, train=False):
...
-        ptsl.eval_conditional_probabilities(input_corpus,
-                                            args.verbose,
-                                            args.train,
-                                            output_file,
-                                            output_proj_file)
+        input_tier = args.input_file[0]
+        input_factors = args.input_file[1]
+        input_corpus = args.input_file[2]
+
+        if not args.output_file:
```
Comment (connormayer) [inline src/pTSL.py]: It might make this simpler if you split this into two separate arguments, `--output_file` and `--verbose_file` or something like that?

Comment (connormayer) [reply src/pTSL.py]: Maybe not though, if you think this is better feel free to keep it this way.

Comment (connormayer) [reply src/pTSL.py]: Alternatively, you can change the default of `output_file` to `None`. I think if you do this you'll get something like this:

```
if output_file is None:
    # No flag
elif len(output_file) == 0:
    # Flag but no argument
elif len(output_file) == 1:
    # Flag and one argument
elif len(output_file) == 2:
    # Flag and two arguments
else:
    # Error?
```

**Review by connormayer — CHANGES_REQUESTED** [review]
I think this is pretty close to being done, just a few small things

File: src/TSL.py (hunk 5 above)
Comment (eyezhick) [reply src/TSL.py:158]: I research a little bit and found this explanation on the official documentation:

newline controls how universal newlines mode works (it only applies to text mode). It can be None, '', '\n', '\r', and '\r\n'. It works as follows:

When reading input from the stream, if newline is None, universal newlines mode is enabled. Lines in the input can end in '\n', '\r', or '\r\n', and these are translated into '\n' before being returned to the caller. If it is '', universal newlines mode is enabled, but line endings are returned to the caller untranslated. If it has any of the other legal values, input lines are only terminated by the given string, and the line ending is returned to the caller untranslated.

When writing output to the stream, if newline is None, any '\n' characters written are translated to the system default line separator, os.linesep. If newline is '' or '\n', no translation takes place. If newline is any of the other legal values, any '\n' characters written are translated to the given string.

So either works in our case, but most implementation I saw has it as '' instead of None as a default.

File: src/pTSL.py (hunk 16 above)
Comment (eyezhick) [reply src/pTSL.py]: I went with the `--output_file` and `--verbose_file` option. I think this is much cleaner.

**Review by eyezhick — COMMENTED** [review]
Made changes as requested and fixed other minor bugs

**Review by connormayer — APPROVED** [review]
Looks good!
